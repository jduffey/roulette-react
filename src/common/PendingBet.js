export class PendingBet {
    constructor(betName, betAmount) {
        this.betName = betName;
        this.betAmount = betAmount;
    }
}
