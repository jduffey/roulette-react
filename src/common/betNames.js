export const BET_NAMES = {
    STRAIGHT_UP_0: "StraightUp_0",
    STRAIGHT_UP_00: "StraightUp_00",
    STRAIGHT_UP_1: "StraightUp_1",
    STRAIGHT_UP_2: "StraightUp_2",
    STRAIGHT_UP_3: "StraightUp_3",
    STRAIGHT_UP_4: "StraightUp_4",
    STRAIGHT_UP_5: "StraightUp_5",
    STRAIGHT_UP_6: "StraightUp_6",
    STRAIGHT_UP_7: "StraightUp_7",
    STRAIGHT_UP_8: "StraightUp_8",
    STRAIGHT_UP_9: "StraightUp_9",
    STRAIGHT_UP_10: "StraightUp_10",
    STRAIGHT_UP_11: "StraightUp_11",
    STRAIGHT_UP_12: "StraightUp_12",
    STRAIGHT_UP_13: "StraightUp_13",
    STRAIGHT_UP_14: "StraightUp_14",
    STRAIGHT_UP_15: "StraightUp_15",
    STRAIGHT_UP_16: "StraightUp_16",
    STRAIGHT_UP_17: "StraightUp_17",
    STRAIGHT_UP_18: "StraightUp_18",
    STRAIGHT_UP_19: "StraightUp_19",
    STRAIGHT_UP_20: "StraightUp_20",
    STRAIGHT_UP_21: "StraightUp_21",
    STRAIGHT_UP_22: "StraightUp_22",
    STRAIGHT_UP_23: "StraightUp_23",
    STRAIGHT_UP_24: "StraightUp_24",
    STRAIGHT_UP_25: "StraightUp_25",
    STRAIGHT_UP_26: "StraightUp_26",
    STRAIGHT_UP_27: "StraightUp_27",
    STRAIGHT_UP_28: "StraightUp_28",
    STRAIGHT_UP_29: "StraightUp_29",
    STRAIGHT_UP_30: "StraightUp_30",
    STRAIGHT_UP_31: "StraightUp_31",
    STRAIGHT_UP_32: "StraightUp_32",
    STRAIGHT_UP_33: "StraightUp_33",
    STRAIGHT_UP_34: "StraightUp_34",
    STRAIGHT_UP_35: "StraightUp_35",
    STRAIGHT_UP_36: "StraightUp_36",
    FIRST_DOZEN: "1st 12",
    SECOND_DOZEN: "2nd 12",
    THIRD_DOZEN: "3rd 12",
    FIRST_18: "1 to 18",
    SECOND_18: "19 to 36",
    EVEN: "Even",
    ODD: "Odd",
    FIRST_COLUMN: "1st Column",
    SECOND_COLUMN: "2nd Column",
    THIRD_COLUMN: "3rd Column",
    RED: "Red",
    BLACK: "Black",
};
