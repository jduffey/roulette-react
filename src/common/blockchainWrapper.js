import { ethers } from "ethers";

const provider = new ethers.providers.JsonRpcProvider("http://localhost:8545");

// TODO: retrieve addresses from file generated by initialization script so they aren't hardcoded
const FIRST_PLAYER_ADDRESS = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266";
const SECOND_PLAYER_ADDRESS = "0x70997970C51812dc3A010C7d01b50e0d17dc79C8";
const THIRD_PLAYER_ADDRESS = "0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC";
const HOUSE_ADDRESS = "0x90F79bf6EB2c4f870365E785982E1f101E93b906";
// Contract addresses are calculated from a hash of the deployer's address and nonce.
const TOKEN_CONTRACT_ADDRESS = "0x057ef64E23666F000b34aE31332854aCBd1c8544";
const ROULETTE_CONTRACT_ADDRESS = "0x261D8c5e9742e6f7f1076Fa1F560894524e19cad";

async function incrementTotalSpins() {
    const signer = provider.getSigner(HOUSE_ADDRESS);
    const incrementTotalSpinsAbi = new ethers.Contract(
        ROULETTE_CONTRACT_ADDRESS,
        ["function incrementTotalSpins()"],
        signer
    );
    const tx = await incrementTotalSpinsAbi.incrementTotalSpins();
    return tx;
}

async function getTotalSpins() {
    const signer = provider.getSigner(HOUSE_ADDRESS);
    const getTotalSpinsAbi = new ethers.Contract(
        ROULETTE_CONTRACT_ADDRESS,
        ["function getTotalSpins() public view returns (uint256)"],
        signer
    );
    const count = await getTotalSpinsAbi.getTotalSpins();
    return count;
}

async function getEthBalance(address) {
    const signer = provider.getSigner(address);
    const balance = await signer.getBalance();
    return ethers.utils.formatEther(balance);
}

// TODO leaving this for reference
// async function sendEther(provider, from, to, amount) {
//     const signer = provider.getSigner(from);
//     const tx = await signer.sendTransaction({
//         to: to,
//         value: ethers.utils.parseEther(amount)
//     });
//     return tx;
// }

async function getBlock() {
    const block = await provider.getBlock();
    return block;
}

async function getTokenBalance(address) {
    const signer = provider.getSigner(address);
    const token = new ethers.Contract(
        TOKEN_CONTRACT_ADDRESS,
        ["function balanceOf(address) view returns (uint)"],
        signer
    );
    const balance = await token.balanceOf(address);
    const formattedBalance = ethers.utils.formatEther(balance);
    return formattedBalance;
}

async function depositEthForTokens(from, amount) {
    const signer = provider.getSigner(from);
    const token = new ethers.Contract(
        TOKEN_CONTRACT_ADDRESS,
        ["function deposit() payable"],
        signer
    );
    const tx = await token.deposit({
        value: ethers.utils.parseEther(amount)
    });
    return tx;
}

async function redeemTokensForEth(from, amount) {
    const signer = provider.getSigner(from);
    const token = new ethers.Contract(
        TOKEN_CONTRACT_ADDRESS,
        ["function withdraw(uint)"],
        signer
    );
    const tx = await token.withdraw(ethers.utils.parseEther(amount));
    return tx;
}

async function transferFrom(from, to, amount) {
    const signer = provider.getSigner(from);
    const token = new ethers.Contract(
        TOKEN_CONTRACT_ADDRESS,
        ["function transferFrom(address, address, uint) returns (bool)"],
        signer
    );
    const tx = await token.transferFrom(from, to, ethers.utils.parseEther(amount));
    return tx;
}

async function incrementJackpotBalance(amount) {
    const signer = provider.getSigner(HOUSE_ADDRESS);
    const contract = new ethers.Contract(
        ROULETTE_CONTRACT_ADDRESS,
        ["function incrementJackpotBalance(uint256)"],
        signer
    );
    const tx = await contract.incrementJackpotBalance(ethers.utils.parseEther(amount));
    return tx;
}

async function getJackpotBalance() {
    const signer = provider.getSigner(HOUSE_ADDRESS);
    const contract = new ethers.Contract(
        ROULETTE_CONTRACT_ADDRESS,
        ["function getJackpotBalance() public view returns (uint256)"],
        signer
    );
    const balance = await contract.getJackpotBalance();
    const formattedBalance = ethers.utils.formatEther(balance);
    return formattedBalance;
}

let tokenSymbol;
(new ethers.Contract(
    TOKEN_CONTRACT_ADDRESS,
    ["function symbol() view returns (string)"],
    provider
)).symbol().then((symbol) => {
    tokenSymbol = symbol;
});

export {
    getEthBalance,
    getBlock,
    getTokenBalance,
    depositEthForTokens,
    redeemTokensForEth,
    transferFrom,
    incrementTotalSpins,
    getTotalSpins,
    incrementJackpotBalance,
    getJackpotBalance,
    FIRST_PLAYER_ADDRESS,
    SECOND_PLAYER_ADDRESS,
    THIRD_PLAYER_ADDRESS,
    HOUSE_ADDRESS,
    TOKEN_CONTRACT_ADDRESS,
    tokenSymbol,
};