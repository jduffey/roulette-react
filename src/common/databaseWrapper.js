const DATABASE_URL = new URL('/player', 'http://localhost:3001');

async function fetchPlayerBalance() {
    const res = await fetch(DATABASE_URL);
    if (!res.ok) {
        console.log("Not OK response from json-server: ", res);
        return `Error: ${res.status} ${res.statusText}`;
    }
    const json = await res.json();
    return json.balance;
}

async function updatePlayerBalance(balance) {
    const res = await fetch(DATABASE_URL, {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({ balance }),
    });
    if (!res.ok) {
        console.log("Not OK response from json-server: ", res);
    }
}

export { fetchPlayerBalance, updatePlayerBalance };
