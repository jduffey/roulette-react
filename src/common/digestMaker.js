// https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest
export async function digestMessage(message) {
    const msgUint8 = new TextEncoder().encode(message);

    const hashBuffer = await crypto.subtle.digest('SHA-256', msgUint8);

    const hashArray = Array.from(new Uint8Array(hashBuffer));

    const hashHex = hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');

    return hashHex;
}
