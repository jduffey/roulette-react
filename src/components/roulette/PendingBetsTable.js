const betRow = (betOption, betAmount) => {
    return (
        <tr key={betOption}>
            <td>
                {betOption}
            </td>
            <td className="pending-bets-info-table-bet-amount">
                {`$ ${betAmount}`}
            </td>
        </tr>
    );
}

const CLASS_NAME = "PendingBetsTable-component";
export function PendingBetsTable(props) {
    const tableBody = (
        <tbody>
            <tr style={{
                textAlign: "center",
            }}>
                <th>Bet Name</th>
                <th>Bet Amount</th>
            </tr>
            {Object.entries(props.pendingBets.reduce((acc, pendingBet) => {
                acc[pendingBet.betName] = (acc[pendingBet.betName] || 0) + pendingBet.betAmount;
                return acc;
            }, {})).map(([betOption, betAmount]) => betRow(betOption, betAmount))}
        </tbody>
    )

    return (
        <div className={CLASS_NAME}>
            <div className="pending-bets-info-table-title-text">
                Pending Bets
            </div>
            <table className="pending-bets-info-table">
                {tableBody}
            </table>
        </div>
    );
}
