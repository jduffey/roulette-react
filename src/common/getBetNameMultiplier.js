import { BET_NAMES } from "./betNames";

export const getBetNameMultiplier = (betName) => {
    console.log("getBetNameMultiplier should be called once per bet placed", betName);
    switch (betName) {
        case BET_NAMES.STRAIGHT_UP_0:
        case BET_NAMES.STRAIGHT_UP_00:
        case BET_NAMES.STRAIGHT_UP_1:
        case BET_NAMES.STRAIGHT_UP_2:
        case BET_NAMES.STRAIGHT_UP_3:
        case BET_NAMES.STRAIGHT_UP_4:
        case BET_NAMES.STRAIGHT_UP_5:
        case BET_NAMES.STRAIGHT_UP_6:
        case BET_NAMES.STRAIGHT_UP_7:
        case BET_NAMES.STRAIGHT_UP_8:
        case BET_NAMES.STRAIGHT_UP_9:
        case BET_NAMES.STRAIGHT_UP_10:
        case BET_NAMES.STRAIGHT_UP_11:
        case BET_NAMES.STRAIGHT_UP_12:
        case BET_NAMES.STRAIGHT_UP_13:
        case BET_NAMES.STRAIGHT_UP_14:
        case BET_NAMES.STRAIGHT_UP_15:
        case BET_NAMES.STRAIGHT_UP_16:
        case BET_NAMES.STRAIGHT_UP_17:
        case BET_NAMES.STRAIGHT_UP_18:
        case BET_NAMES.STRAIGHT_UP_19:
        case BET_NAMES.STRAIGHT_UP_20:
        case BET_NAMES.STRAIGHT_UP_21:
        case BET_NAMES.STRAIGHT_UP_22:
        case BET_NAMES.STRAIGHT_UP_23:
        case BET_NAMES.STRAIGHT_UP_24:
        case BET_NAMES.STRAIGHT_UP_25:
        case BET_NAMES.STRAIGHT_UP_26:
        case BET_NAMES.STRAIGHT_UP_27:
        case BET_NAMES.STRAIGHT_UP_28:
        case BET_NAMES.STRAIGHT_UP_29:
        case BET_NAMES.STRAIGHT_UP_30:
        case BET_NAMES.STRAIGHT_UP_31:
        case BET_NAMES.STRAIGHT_UP_32:
        case BET_NAMES.STRAIGHT_UP_33:
        case BET_NAMES.STRAIGHT_UP_34:
        case BET_NAMES.STRAIGHT_UP_35:
        case BET_NAMES.STRAIGHT_UP_36:
            return 35;

        case BET_NAMES.FIRST_DOZEN:
        case BET_NAMES.SECOND_DOZEN:
        case BET_NAMES.THIRD_DOZEN:
        case BET_NAMES.FIRST_COLUMN:
        case BET_NAMES.SECOND_COLUMN:
        case BET_NAMES.THIRD_COLUMN:
            return 2;

        case BET_NAMES.FIRST_18:
        case BET_NAMES.SECOND_18:
        case BET_NAMES.EVEN:
        case BET_NAMES.ODD:
        case BET_NAMES.RED:
        case BET_NAMES.BLACK:
            return 1;

        default:
            throw new Error("Invalid bet name", betName);
    }
}
