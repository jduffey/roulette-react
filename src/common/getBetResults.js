import { getBetNameMultiplier } from "../common/getBetNameMultiplier";
import { getWinningCriteria } from "../common/getWinningCriteria";

export function getBetResults(startingBalance, bets, winningWheelNumber) {

    const betResults = Object.keys(bets).reduce((acc, betOption) => {
        const multiplier = getBetNameMultiplier(betOption);
        const betAmount = bets[betOption];
        const winningsOnBet = getWinningCriteria(winningWheelNumber).has(betOption) ? betAmount * multiplier : 0;

        acc[betOption] = {
            winnings: winningsOnBet,
            betReturned: winningsOnBet ? betAmount : 0,
        };

        return acc;
    }, {});

    const totalBetAmount = Object.values(bets).reduce((acc, betAmount) => {
        return acc + betAmount;
    }, 0);

    const winningsPlusBetsReturned =
        Object.values(betResults).reduce((acc, betResult) => {
            return acc + betResult.winnings + betResult.betReturned;
        }, 0);

    const result = {
        startingBalance: startingBalance,
        betResults: betResults,
        winningWheelNumber: winningWheelNumber,
        resultingBalance: startingBalance - totalBetAmount + winningsPlusBetsReturned,
    };

    return result;
}
