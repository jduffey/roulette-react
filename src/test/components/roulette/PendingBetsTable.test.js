import renderer from 'react-test-renderer';

import { PendingBetsTable } from '../../../components/roulette/PendingBetsTable';
import { PendingBet } from '../../../common/PendingBet';

describe('PendingBetsTable', () => {
    it('renders with no bets', () => {
        const sut =
            PendingBetsTable({
                pendingBets: [],
            });

        const actual = renderer.create(sut);

        expect(actual).toMatchSnapshot();
    });

    it('renders with bets', () => {
        const sut =
            PendingBetsTable({
                pendingBets: [
                    new PendingBet("foo", 10),
                    new PendingBet("bar", 20),
                    new PendingBet("baz", 30),
                ],
            });

        const actual = renderer.create(sut);

        expect(actual).toMatchSnapshot();
    });
});
