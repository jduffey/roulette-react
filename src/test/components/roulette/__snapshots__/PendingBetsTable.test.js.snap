// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingBetsTable renders with bets 1`] = `
<div
  className="PendingBetsTable-component"
>
  <div
    className="pending-bets-info-table-title-text"
  >
    Pending Bets
  </div>
  <table
    className="pending-bets-info-table"
  >
    <tbody>
      <tr
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <th>
          Bet Name
        </th>
        <th>
          Bet Amount
        </th>
      </tr>
      <tr>
        <td>
          foo
        </td>
        <td
          className="pending-bets-info-table-bet-amount"
        >
          $ 10
        </td>
      </tr>
      <tr>
        <td>
          bar
        </td>
        <td
          className="pending-bets-info-table-bet-amount"
        >
          $ 20
        </td>
      </tr>
      <tr>
        <td>
          baz
        </td>
        <td
          className="pending-bets-info-table-bet-amount"
        >
          $ 30
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`PendingBetsTable renders with no bets 1`] = `
<div
  className="PendingBetsTable-component"
>
  <div
    className="pending-bets-info-table-title-text"
  >
    Pending Bets
  </div>
  <table
    className="pending-bets-info-table"
  >
    <tbody>
      <tr
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <th>
          Bet Name
        </th>
        <th>
          Bet Amount
        </th>
      </tr>
    </tbody>
  </table>
</div>
`;
